---
title: "Dendrobates_auratus_transcriptome"
author: "Adam Stuckert"
date: '`r format(Sys.time(), "%d %B, %Y")`'
output:
  html_document: default
  pdf_document: default
---

```{r global_options, include=FALSE}
knitr::opts_chunk$set(echo=FALSE, warning=FALSE, message=FALSE)
library(biomaRt)
library(sleuth)
library(dplyr)
library(foreach)
library(doParallel)
library(data.table)
library(ggthemes)
library(scales)
library(gridExtra)
```

Print R information so that I can easily pick out information for publication.

```{r}
sessionInfo()
```

```{r}

setwd("~/R/R projects/AuratusTranscriptome")
base_dir <- "C:/Users/Adam Stuckert/Documents/R/R projects/AuratusTranscriptome"
sample_id <- dir(file.path(base_dir, "kallisto_quants"))
#sample_id
kal_dirs <- sapply(sample_id, function(id) file.path(base_dir, "kallisto_quants", id))
#kal_dirs
samples <- read.table("auratus_samples.txt", header = TRUE)
samples <- samples[order(samples$sample),]
samples <- dplyr::mutate(samples, path = kal_dirs)
samples
```



```{r}
so <- sleuth_prep(samples) #, transformation_function = function(x) log2(x + 0.5)

plot_pca(so, text_labels = TRUE)
```


I can skip the above chunk of code and come here in the future, and just load the sleuth object (so <- sleuth_load(auratus_sleuth))

```{r}
plot_pca(so, point_size = 0.001, units = 'est_counts') + geom_point() + aes(size = 3, colour = samples$morph) + scale_colour_manual(values = c("navyblue","forestgreen","tan3","dodgerblue"), guide = guide_legend(values=c("navyblue","forestgreen","tan3","dodgerblue"), title = "Morph", override.aes = list(size=4))) + guides(size=FALSE)

ggsave("auratus_est-counts_pca.tiff", width = 6.81, height = 3.99)
```


```{r}
plot_pca(so, point_size = 0.001, units = 'tpm') + geom_point() + aes(size = 3, colour = samples$morph) + scale_colour_manual(values = c("navyblue","forestgreen","tan3","dodgerblue"), guide = guide_legend(values=c("navyblue","forestgreen","tan3","dodgerblue"), title = "Morph", override.aes = list(size=4))) + guides(size=FALSE)

ggsave("auratus_tpm_pca.tiff", width = 6.81, height = 3.99)
```




```{r}
plot_loadings(so, pc_input = 1)
plot_loadings(so, pc_input = 2)
```

Now I'll annotation files to look at overall gene expression and finally do a gene ontology analysis. We have collated a list of genes involved in color and pattern (largely from zebrafish studies) which we will target. I've downloaded the *Xenopus* and *Nanorana* peptide databases, and also downloaded our list of proteins from the Uniref database.

```{r, eval=FALSE}
# this is to download the xenopus data from biomart. don't run after the first time.
ensembl=useMart("ensembl")
 xenopus <- useMart("ensembl", dataset = "xtropicalis_gene_ensembl")

 t2g <- biomaRt::getBM(attributes = c("ensembl_peptide_id_version",  "external_gene_name"), mart = xenopus)    
 write.table(t2g, "xenpep2gene.tsv", sep = "\t")
```


```{r}

# First download the  annotation from linux/enrique
ann <- read.table("newallpep_tophit.txt", header = FALSE, fill = TRUE)
colnames(ann)[c(1:12)] <- c("transcript_id", "peptide_id", "percentage_id_matches",
                              "alignment_length", "number_mismatches", "number_gap_openings",
                              "query_start", "query_end", "alignment_start", "alignment_end",
                              "expected_value", "bitscore")

### Load the annotation data
# Xenopus
x2g <- fread("annotationmappingdocuments/xenpep2gene.tsv", header = FALSE)
x2g <- x2g[,-1]
colnames(x2g) <- c("peptide_id", "gene_name")

# Nanorana
n2g <- fread("annotationmappingdocuments/nan2gene.tsv", header = FALSE)
colnames(n2g) <- c("peptide_id", "gene_name")

# Uniref
u2g <- fread("annotationmappingdocuments/uniprotlistnew", header = TRUE)
u2g <- u2g[,c(1,5)]
u2g$peptide_id <- ""
u2g$peptide_id <- paste("UniRef90_", u2g$Entry)
u2g <- u2g[,c(3,2)]
colnames(u2g) <- c("peptide_id", "gene_name")

# merge all
a2g <- rbind(x2g, n2g, u2g)

a2g$gene_name <- tolower(a2g$gene_name)

xen <- ann[,c(1:2)]


biggie <- dplyr::left_join(xen, a2g, by = "peptide_id")
colnames(biggie)[1] <- "target_id"

pop_design <- model.matrix(~0 + samples$morph)
colnames(pop_design) <- levels(samples$morph)



sopop <- sleuth_prep(samples, full_model = pop_design, num_cores = 6, target_mapping = biggie)

sopop <- sleuth_fit(sopop, formula = pop_design, fit_name = "full")
sopop <- sleuth_fit(sopop, formula = ~ 1, fit_name = "reduced")
#so <- sleuth_fit(so, full_model = spline_design)
models(sopop)
sopop_lrt <- sleuth_lrt(sopop, "reduced", "full")
# so_wt <- sleuth_wt(so, "reduced", "full")
# Ok I don't know why that doesn't actually work....

pop_lrt_results <- sleuth_results(sopop_lrt, 'reduced:full', test_type = 'lrt')
table(pop_lrt_results[,"qval"] < 0.05) #2,768 significant transcripts between the null and the others...
pop_sig_results <-  pop_lrt_results[order(pop_lrt_results$qval),]
pop_siggies <- pop_sig_results[ which(pop_sig_results$qval < 0.05),]
write.csv(pop_siggies, "auratus_significant_genes.csv")
```



Ok, 2,768 'significant' transcripts (statistically speaking). This is about 8.1% of transcripts. Now on to do some differential expression by candidate color genes, followed by some fancy shit like GO analyses. Now is probably a good time to attempt to learn that bullshit.



I will now attempt to get the loadings from the PCA, which uses the function prcomp().

```{r}

spread_abundance_by <- function(abund, var, which_order) {
  # var <- lazyeval::lazy(var)
  var_spread <- abund %>%
    select_("target_id", "sample", var) %>%
    tidyr::spread_("sample", var) %>%
    as.data.frame(stringsAsFactors = FALSE)

  rownames(var_spread) <- var_spread$target_id
  var_spread["target_id"] <- NULL

  result <- as.matrix(var_spread)

  result[, which_order, drop = FALSE]
}

mat <- spread_abundance_by(so$obs_norm, "tpm",
      so$sample_to_covariates$sample)

#this will run the pca, and this is what is actually being run in sleuth
 pca_res <- prcomp(t(mat))
 
 # This gives the relative contribution of each principal component, as well as the std
 summary(pca_res) 
 
 # The following bits probably aren't needed, but may be interesting for looking at contributions of specific transcripts (ie, color genes)
 pca_res$rotation %>% head
 
 aload <- abs(pca_res$rotation) #saves the absolute values
 
 # this gives the relative contribution to each principal component...
 sweep(aload, 2, colSums(aload), "/") %>% head 
 
 # For example, Melanocortin receptor is probably Transcript_23342, so I can look at that transcript
 # cont <- sweep(aload, 2, colSums(aload), "/") 
 # cont <- as.data.frame(cont)
 # cont["Transcript_23342",]
 
 # OK, soooo not a big contribution here.
 
```


Looks like PC1 = 37% of the variance and PC2 = 21%. Doesn't explain a huge amount of the variation.


Load the importance for both PC1 and PC2 in descending order (ie, those with the biggest effect on the principal component will be first).


```{r}
pc <- aload[,c(1:2)]
pc <- as.data.frame(pc)
pc <- tibble::rownames_to_column(pc, var = "transcript")

pc1 <- pc[order(-pc$PC1),]
pc1 <- pc1[,-3]

pc2 <- pc[order(-pc$PC2),]
pc2 <- pc2[,-2]


```





Get the significant ones, and only those that are classified as "genes" with an external gene name. I'm going to load this into GOrilla for now.

```{R}
GO <- dplyr::filter(pop_siggies, gene_name != "NA")
GO <- dplyr::filter(GO, gene_name != "")

write.csv(GO, "all_peptides_genes4GOrilla.csv")
```


Now I will go through the series of candidate color genes to see what is differentially expressed between the color morphs. These color genes are saved in a csv which we've compiled.


```{r}
colors <- read.csv("color_genes.csv")
colnames(colors)[1] <- "gene_name"


# Make them all lower case...
colors$gene_name <- tolower(colors$gene_name)

pop_lrt_results$gene_name <- tolower(pop_lrt_results$gene_name)
colors <- dplyr::left_join(colors, pop_lrt_results, by = "gene_name")

## Order them by 'significance'
colors <-  colors[order(colors$gene_name),]

# Drop color genes that don't have an associated transcript
colors <- dplyr::filter(colors, target_id != "NA")

## These are just the 'statistically significant' hits
sig_colors <- dplyr::filter(colors, qval < 0.05)



##### This will iterate through and make a nice figure for each of the statistically significant color genes
for (i in 1:nrow(sig_colors)){
  transcript <- sig_colors[i,"target_id"]
  gene <- sig_colors[i,"gene_name"]
  qval <- sig_colors[i,"qval"]
  tmp <- sopop$obs_norm %>% dplyr::filter(target_id == transcript)  ### These are normalized values I think!
  tmp <- dplyr::full_join(sopop$sample_to_covariates, tmp, by = 'sample')
  tmp
  
  
a <- ggplot(tmp, aes(x=morph, y=tpm)) + xlab("Morph") + ylab("Transcripts per million") + geom_jitter(aes(size = 3, color = morph), width = 0.2, height = 0) + scale_colour_manual(values = c("navyblue","forestgreen","tan3","dodgerblue"), guide = guide_legend(title = "Color morph", override.aes = list(size=4))) +  geom_smooth(method = loess) + ggtitle(paste0(gene, ", ", "\n(", transcript, "), q value = ", qval)) + guides(size=FALSE) + theme_bw() + theme(panel.grid.major = element_blank())

 
  ggsave(paste0("colorgene-figures/", gene, "-", transcript, ".png"), width = 6.81, height = 3.99)
}


write.csv(colors, "auratus_color_gene_stats.csv")
write.csv(sig_colors, "auratus_significant_color_gene_stats.csv")

```

Extract fold changes.

```{r}
sopop <- sleuth_wt(sopop, 'Blue-Black')
sopop <- sleuth_wt(sopop, 'Microspot')
sopop <- sleuth_wt(sopop, 'San-Felix')
sopop <- sleuth_wt(sopop, 'Super-Blue')
bbwalds <- sleuth_results(sopop, 'Blue-Black', test_type = "wt", which_model = "full",
  rename_cols = TRUE, show_all = TRUE)
bb2bb<- sleuth_results(sopop, 'Blue-Black', test_type = "wt", which_model = "full",
  rename_cols = TRUE, show_all = TRUE)
bb2micro <- sleuth_results(sopop, 'Microspot', test_type = "wt", which_model = "full",
  rename_cols = TRUE, show_all = TRUE)
bb2sanfel <- sleuth_results(sopop, 'San-Felix', test_type = "wt", which_model = "full",
  rename_cols = TRUE, show_all = TRUE)
bb2super <- sleuth_results(sopop, 'Super-Blue', test_type = "wt", which_model = "full",
  rename_cols = TRUE, show_all = TRUE)


pop_design <- model.matrix(~0 + samples$morph)
colnames(pop_design) <- levels(samples$morph)

micro_pop_design <- pop_design[,c(2,3,4,1)]
sanfel_pop_design <- pop_design[,c(3,4,1,2)]



sopop <- sleuth_fit(sopop, formula = micro_pop_design, fit_name = "2full")
sopop <- sleuth_fit(sopop, formula = sanfel_pop_design, fit_name = "3full")

sopop <- sleuth_wt(sopop, 'Microspot', '2full')
sopop <- sleuth_wt(sopop, 'San-Felix', '2full')
sopop <- sleuth_wt(sopop, 'Super-Blue', '2full')
micro2sanfel <- sleuth_results(sopop, 'San-Felix', test_type = "wt", which_model = "2full",
  rename_cols = TRUE, show_all = TRUE)
micro2super <- sleuth_results(sopop, 'Super-Blue', test_type = "wt", which_model = "2full",
  rename_cols = TRUE, show_all = TRUE)

```



Extract effect sizes...

```{r}

# This will get the effect sizes for the morphs by gene....
full_model <- extract_model(sopop, 'full')

morphs <- list('Blue-Black', 'Microspot', 'San-Felix', 'Super-Blue')

size <- nrow(full_model)

effect_sizes <- data.frame(target_id = unique(full_model$target_id))

for (i in morphs){
  tmp <- dplyr::filter(full_model, term == i)
  tmp <- tmp[,c(1,3,4)]
  colnames(tmp)[2] <- paste0(i, "_estimate")
  colnames(tmp)[3] <- paste0(i, "_std_error")
  effect_sizes <- merge(effect_sizes, tmp, by = "target_id")
}


# get est? idk man.
kaltab <- kallisto_table(sopop)

eh <- kaltab %>% group_by(target_id, morph) %>% summarise(counts = mean(est_counts)) 
# get unique rows????

eh2 <- kaltab %>% group_by(target_id, morph) %>% summarise(error = sd(est_counts))

eh3 <- dplyr::bind_cols(eh,eh2)



###left off here, quick for loop to make shit happen with combining columns or some shit.





iris %>%
 group_by(Species) %>%
 summarise(avg = mean(Sepal.Width)) %>%
 arrange(avg)

effect_sizes <- dplyr::inner_join(
  dplyr::select(dplyr::filter(full_model, term == 'Blue-Black'),
    target_id, full_estimate = estimate),
  dplyr::select(dplyr::filter(reduced_model, term == 'Blue-Black'),
    target_id, reduced_estimate = estimate),
  by = 'target_id'
  )

dplyr::inner_join(
  dplyr::select(top_different[1:10, ], target_id),
  effect_sizes,
  by = 'target_id')

```




We ended up losing some of the color specific genes, I assume because there was a *slightly better* mapping to either *Rana* or *Nanorana*, so we will do the same as above, but just with the *Xenopus* annotation to try and recover those color genes. It seems like there could be a way to combine these programmatically, but this is just a first pass.


```{r}
colors <- read.csv("color_genes.csv")
colnames(colors)[1] <- "gene_name"

# First download the Xenopus annotation from linux/enrique
xenann <- read.table("newxen_tophit.txt", header = FALSE, fill = TRUE)
colnames(xenann)[c(1:12)] <- c("transcript_id", "peptide_id", "percentage_id_matches",
                              "alignment_length", "number_mismatches", "number_gap_openings",
                              "query_start", "query_end", "alignment_start", "alignment_end",
                              "expected_value", "bitscore")


xenopus <- xenann[,c(1:2)]


xen_full <- dplyr::left_join(xenopus, x2g, by = "peptide_id")

xen_full <- dplyr::rename(xen_full, target_id = transcript_id)

pop_design <- model.matrix(~0 + samples$morph)
colnames(pop_design) <- levels(samples$morph)



sopop_b <- sleuth_prep(samples, full_model = pop_design, num_cores = 6, target_mapping = xen_full)

sopop_b <- sleuth_fit(sopop_b, formula = pop_design, fit_name = "full")
sopop_b <- sleuth_fit(sopop_b, formula = ~ 1, fit_name = "reduced")
#so <- sleuth_fit(so, full_model = spline_design)
models(sopop_b)
sopop_b_lrt <- sleuth_lrt(sopop_b, "reduced", "full")
# so_wt <- sleuth_wt(so, "reduced", "full")
# Ok I don't know why that doesn't actually work....

pop_b_lrt_results <- sleuth_results(sopop_b_lrt, 'reduced:full', test_type = 'lrt')
table(pop_b_lrt_results[,"qval"] < 0.05) #2,824 significant transcripts between the null and the others...
pop_b_sig_results <-  pop_b_lrt_results[order(pop_b_lrt_results$qval),]
pop_b_siggies <- pop_b_sig_results[ which(pop_b_sig_results$qval < 0.05),]

write.csv(pop_b_siggies, "auratus_xenopus_significant_genes.csv")

GOb <- dplyr::filter(pop_b_siggies, gene_name != "NA")
GOb <- dplyr::filter(GOb, gene_name != "")

write.csv(GOb, "xenopus_only_genes4GOrilla.csv")



pop_b_lrt_results$gene_name <- tolower(pop_b_lrt_results$gene_name)

colorsb <- dplyr::left_join(colors, pop_b_lrt_results, by = "gene_name")

## Order them by 'significance'
colorsb <-  colorsb[order(colorsb$qval),]

# Drop color genes that don't have an associated transcript
colorsb <- dplyr::filter(colorsb, target_id != "NA")

## These are just the 'statistically significant' hits
sig_b_colors <- dplyr::filter(colorsb, qval < 0.05)

write.csv(colorsb, "auratus_xenopus_color_gene_stats.csv")
write.csv(sig_b_colors, "auratus_xenopus_significant_color_genes.csv")

##### This will iterate through and make a nice figure for each of the statistically significant color genes
for (i in 1:nrow(sig_b_colors)){
  transcript <- sig_b_colors[i,"target_id"]
  gene <- sig_b_colors[i,"gene_name"]
  qval <- sig_b_colors[i,"qval"]
  tmp <- sopop_b$obs_norm %>% dplyr::filter(target_id == transcript)  ### These are normalized values I think!
  tmp <- dplyr::full_join(sopop_b$sample_to_covariates, tmp, by = 'sample')
  tmp
  
  
a <- ggplot(tmp, aes(x=morph, y=tpm)) + xlab("Morph") + ylab("Transcripts per million") + geom_jitter(aes(size = 3, color = morph), width = 0.2, height = 0) + scale_colour_manual(values = c("navyblue","forestgreen","tan3","dodgerblue"), guide = guide_legend(title = "Color morph", override.aes = list(size=4))) +  geom_smooth(method = loess) + ggtitle(paste0(gene, ", ", "\n(", transcript, "), q value = ", qval)) + guides(size=FALSE) + theme_bw() + theme(panel.grid.major = element_blank())

 
  ggsave(paste0("xenopuscolorgene-figures/", gene, "-", transcript, ".png"), width = 6.81, height = 3.99)
}

```


Now, why are there differences between the full annotation and the *Xenopus* annotation?

```{r}
# xen
xencolors <- sig_b_colors[,c(2,1,5, 15)]
colnames(xencolors) <- c("target_id", "xen_gene_name", "qval", "xen_peptide_id")

# full
allcolors <- pop_lrt_results[,c(1,15,14)]
colnames(allcolors) <- c("target_id", "all_gene_name", "all_peptide_id")

# get blast values, ann = full annotation, xenann = xenopus annotation
# full first
ann1 <- ann[,c(1,2,11)]
colnames(ann1) <- c("target_id", "all_peptide-blast", "all_evalue")
#Xenopus second
xenann1 <- xenann[,c(1,2,11)]
colnames(xenann1) <- c("target_id", "xen_peptide-blast", "xen_evalue")

# merge
xentable <- dplyr::left_join(xencolors, xenann1, by = "target_id")
alltable <- dplyr::left_join(ann1, allcolors, by = "target_id")

table <- dplyr::left_join(xentable, alltable, by = "target_id")
table <- table[order(table$xen_gene_name),]
write.csv(table, "xenvsall_evals_colorgenes.csv")

## now full 2 xen
all <- sig_colors[,c(2,1,5,15)]
colnames(all) <- c("target_id", "all_gene_name", "qval",  "all_peptide_id")

# xen
xentran <- pop_b_lrt_results[,c(1,15,14)]
colnames(xentran) <- c("target_id", "xen_gene_name", "xen_peptide_id")

# merge
#alltable <- dplyr::left_join(all, alltable, by = "target_id")
table2 <- dplyr::left_join(alltable, xentable, by = "target_id")
table2 <- table2[order(table2$all_gene_name)]
write.csv(table2, "allvsxen_evals_colorgenes.csv")


pop_lrt_results[order(pop_lrt_results$qval),]

realtable <- merge(all, xentable, by = "target_id", all = TRUE)
realtable <- realtable[order(realtable$all_gene_name),]

write.csv(realtable, "annotation_values_colorgenes.csv")
```


Make some figures from Panther:

```{r}
bio <- read.csv("panther/panther-biologicalprocesses.csv")
cell <- read.csv("panther/panther-cellularcomponent.csv")
mol <- read.csv("panther/panther-molecularfunction.csv")

# make a theme
pietheme <- theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank())

colourCountb = length(unique(bio$Category_name))
colourCountc = length(unique(cell$Category_name))
colourCountm = length(unique(mol$Category_name))
getPalette = colorRampPalette(brewer_pal(palette = "Set1")(9))


bp <- ggplot(bio, aes(x="", y=Genes, fill=Category_name)) + geom_bar(width = 1, stat = "identity")
bp <- bp + coord_polar("y") + theme_minimal() + ylab("Biological processes") + xlab("") + labs(fill="Category")  + pietheme +   scale_fill_manual(values = getPalette(colourCountb)) 
bp
ggsave("panther/biologicalprocesses-pie.png", width = 6.81, height = 3.99, dpi = 600)


bp1 <- ggplot(bio, aes(x="", y=Genes, fill=Category_name)) + geom_bar(width = 1, stat = "identity")
bp1 <- bp1 + theme_minimal() + ylab("Biological processes") + xlab("") + labs(fill="Category")  + pietheme +  scale_fill_canva(aes(x="", y=Genes, fill=Category_name), palette = "Cool blues") 
bp1
ggsave("panther/biologicalprocesses-bar.png", width = 6.81, height = 3.99, dpi = 600)

# Cellular processes pie chart
bp2 <- ggplot(cell, aes(x="", y=Genes, fill=Category_name)) + geom_bar(width = 1, stat = "identity")
bp2 <- bp2 + coord_polar("y") + theme_minimal() + ylab("Cellular component") + xlab("") + labs(fill="Category")  + pietheme +   scale_fill_manual(values = getPalette(colourCountc))
bp2
ggsave("panther/cellularcomponent-pie.png", width = 6.81, height = 3.99, dpi = 600)

# molecular function pie chart
bp3 <- ggplot(mol, aes(x="", y=Genes, fill=Category_name)) + geom_bar(width = 1, stat = "identity")
bp3 <- bp3 + coord_polar("y") + theme_minimal() + ylab("Molecular function") + xlab("") + labs(fill="Category")  + pietheme +   scale_fill_manual(values = getPalette(colourCountm))
bp3
ggsave("panther/molecularfunction-pie.png", width = 6.81, height = 3.99, dpi = 600)



# Produce the same plots as above, but just make them bar charts
# Biological function
# bio <-  bio[order(-bio$Genes),]
bio$Category_name <- factor(bio$Category_name, levels = bio$Category_name[order(bio$Genes)])
bp4 <- ggplot(bio, aes(x=Category_name, y=Genes)) + geom_bar(width = 1, stat = "identity") 
bp4 <- bp4 + coord_flip() + ylab("Number of genes") + xlab("Biological processes") + labs(fill="Category") + theme_bw()   
bp4
ggsave("panther/biologicalprocesses-bar.png", width = 6.81, height = 3.99, dpi = 600)

# Cellular processes
cell$Category_name <- factor(cell$Category_name, levels = cell$Category_name[order(cell$Genes)])
bp5 <- ggplot(cell, aes(x=Category_name, y=Genes)) + geom_bar(width = 1, stat = "identity")
bp5 <- bp5 + coord_flip() + ylab("Number of genes") + xlab("Cellular processes") + labs(fill="Category") + theme_bw()   
bp5
ggsave("panther/Cellularprocesses-bar.png", width = 6.81, height = 3.99, dpi = 600)

# Molecular function
mol$Category_name <- factor(mol$Category_name, levels = mol$Category_name[order(mol$Genes)])
bp6 <- ggplot(mol, aes(x=Category_name, y=Genes)) + geom_bar(width = 1, stat = "identity")
bp6 <- bp6 + coord_flip() + ylab("Number of genes") + xlab("Molecular function") + labs(fill="Category") + theme_bw()   
bp6
ggsave("panther/molecularfunction-bar.png", width = 6.81, height = 3.99, dpi = 600)

# Combine the three boxplots
com <- grid.arrange(bp4, bp5, bp6, nrow = 3)
com
ggsave("panther/combinedbarplots.png", com, width = 6.81, height = 12, dpi = 600)
````

